Feature: prev next button for detail and edit views (#9165)
marmelab/react-admin: docs/usePrevNextController.md

---
layout: default
title: "usePrevNextController"
---

# `usePrevNextController`

`useShowController` fetche previous and next record IDs for a given record and resource.

It fetches the list of records from the REST API according to the filters and the sort order of listsfrom the store and also merges the filters and the sorting order passed into props.

> It used by `<PrevNextButtons>` to render a navigation to move to the next or previous record of a resource.

Reviewer: ```suggestion
It is used by the `<PrevNextButtons>` component to render navigation buttons that allow users to navigate to the next or previous record of a resource.
```

                aria-errormessage={error.message}
            />
        );
    }

    return (
        <nav>
            <ul className={clsx(PrevNextButtonClasses.list)}>
                <li>
                    <IconButton
                        component={hasPrev ? Link : undefined}
                        to={navigateToPrev}
                        aria-label={translate('ra.navigation.previous')}
                        disabled={!hasPrev}
                    >
                        <ArrowBackIcon />
                    </IconButton>
                </li>
                {typeof index === 'number' && (
                    <li>
                        {index + 1} / {total}
                    </li>
                )}
                <li>
                    <IconButton
                        component={hasNext ? Link : undefined}
                        to={navigateToNext}
                        aria-label={translate('ra.navigation.next')}
                        disabled={!hasNext}
                    >
                        <ArrowForwardIcon />
                    </IconButton>
                </li>
            </ul>
        </nav>
    );
};

> interface MyPrevNextButtonProps<RecordType extends RaRecord = any> extends UsePrevNextControllerProps<RecordType> 

Reviewer: ```suggestion
interface MyPrevNextButtonProps<RecordType extends RaRecord = any> extends UsePrevNextControllerProps<RecordType> {}
```

| `limit`        | Optional | `number`         | `1000`                              | Maximum number of records to fetch.                                                         |
| `linkType`     | Optional | `string`         | 'edit'                              | Specifies the view to redirect to when navigating.                                          |
| `queryOptions` | Optional | `object`         | `{ staleTime: 5 * 60 * 1000 }`      | The options to pass to the useQuery hook.                                                   |
| `resource`     | Optional | `string`         | -                                   | The resource name, e.g. `customers`.                                                        |
| `sort`         | Optional | `object`         | `{ field: 'id', order: SORT_ASC } ` | The initial sort parameters.                                                                |
| `storeKey`     | Optional | `string | false` | -                                   | The key to use to match a filter & sort configuration of a `<List>`. Pass false to disable. |

## `filter`

> Just like [Permanent `filter` in `<List>`](./List.md#filter-permanent-filter), you can specify what filter to apply to fetches the list of records.

Reviewer: ```suggestion
Just like [Permanent `filter` in `<List>`](./List.md#filter-permanent-filter), you can specify what filters to apply when fetching the list of records.
```

## `limit`

You can set the maximum number of records to fetch with `limit` prop. By default, `usePrevNextController` fetches a maximum of `1000` records.

## `linkType`

> By default `<usePrevNextController>` generates links to `<Edit>` view. You can also set the `linkType` prop to `show` to link to the `<Show>` view instead.

Reviewer: ```suggestion
By default `<usePrevNextController>` generates links to the `<Edit>` view. You can also set the `linkType` prop to `show` to link to the `<Show>` view instead.
```

`linkType` accepts the following values:

* `linkType="edit"`: links to the edit page. This is the default behavior.
* `linkType="show"`: links to the show page.

## `queryOptions`

`usePrevNextController` accepts a `queryOptions` prop to pass options to the react-query client. 

This can be useful e.g. to pass [a custom `meta`](./Actions.md#meta-parameter) to the `dataProvider.getList()` call.

## `resource`

> By default, `usePrevNextController` operates on the current `ResourceContext` (defined at the routing level), so under the `/customers` path, the `resource` prop will be `customers`. You may want to force a different resource to fetche. In this case, pass a custom `resource` prop, and it will override the `ResourceContext` value.

Reviewer: ```suggestion
By default, `usePrevNextController` operates on the current `ResourceContext` (defined at the routing level), so under the `/customers` path, the `resource` prop will be `customers`. Pass a custom `resource` prop to override the `ResourceContext` value.
```

## `sort`

> Pass an object literal as the `sort` prop to determine `field` and `order` used for sorting:

Reviewer: ```suggestion
Pass an object literal as the `sort` prop to set the `field` and `order` used for sorting:
```

## `storeKey`

> `usePrevNextController` grabs list parameters (sort and filters) from the store.

Reviewer: ```suggestion
`usePrevNextController` gets the list parameters (sort and filters) from the store.
```

> This prop is useful if you specified a custom `storeKey` for a `<List>` and you want `usePrevNextController` grabs the same stored parmaters.

Reviewer: ```suggestion
This prop is useful if you specified a custom `storeKey` for a `<List>` and you want `usePrevNextController` to use the same stored parameters.
```
